Prototype: move blocks between levels with keyboard

tellthemachines opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Addresses #22031

Using the toolbar action described in the issue above to enable moving blocks in and out of nesting levels with keyboard only (something that is currently only possible with drag and drop). 

~~I'm leveraging navigation mode for this, so I had to incidentally address #20732 too, to enable navigating in and out of nesting levels.~~ Update: #20732 has been addressed separately.

~~Setting this up as a draft so we can test the keyboard interaction. It doesn't yet work with a mouse/touchpad/touchscreen, as I'm not sure what the correct behaviour should be for those interaction modes.~~ Update: this is now ready for review. Basic mouse interaction is enabled: clicking on "Move to" and then clicking once on the target block, then pressing Space or Enter, will move the block into place above the target block.

All feedback and testing appreciated!

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

So far, tested only in browser with keyboard. A quick spin with VoiceOver shows the keyboard flow works, but there are no announcements of the actions taken (that should also be implemented once we decide if this is the way we want to go).

How it works:

1. In the toolbar for the block you want to move, click on "Move to";
2. Use the Up and Down Arrow keys to navigate within the same level, use the Left Arrow key to go up one level, and the Right Arrow key to go down one level (when you're on a block that has children);
3. When the blue line is in the desired target position, press Enter or Space to move the block to that position.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

talldan commented:
It's a really promising prototype. 

The improvements that make it possible to select nested blocks (#20732) in navigation mode work really well, and I think it'd be great to see them in a separate PR so that they can be shipped more quickly.

And in terms of the Move To option, I've seen some similar systems for moving items around that work pretty well. Android TV uses a similar system for reordering apps, where an app is _grabbed_, _moved_ and then _dropped_. 

There are a few things I found in testing (though I realise this is currently a prototype, so I understand they might already be part of the thought process):
- I became easily confused over whether to use space or enter to move the block. I feel like both should be allowed.
- It'd be cool if the block actually moved as part of the visualisation.
- Some indication of when it's not possible to drop a block into a location will probably be needed, e.g. moving a paragraph block into `columns` isn't possible, the user has to move to the column block instead. Potentially there shouldn't even be drop zones for columns, though it may also be confusing if moving a block bypasses a level.
- Possibly some indication of when left and right can be used to nest and unnest might be helpful.

enriquesanchez commented:
Hi @tellthemachines 👋

This is really cool! The keyboard interaction feels natural. I too have similar feedback as @talldan's:

- I was expecting `enter` to work, instead of `spacebar`. Maybe allow both?
- On a group block with two child paragraphs inside, it wasn't visually clear to me that the child block had moved outside. I think this is because there was no visual reference to compare to. Maybe if the containing block could also be somehow highlighted, this would be more obvious? 

What I did was reference the block breadcrumbs at the bottom of the editor to be able to tell when the paragraph was outside the group.

Overall, this is really promising! 👍

tellthemachines committed:
Prototype move blocks between levels with keyboard

tellthemachines committed:
Add "Move To" action back

tellthemachines committed:
Allow moving blocks with space or enter.

tellthemachines committed:
Only show drop target when block can be moved.

tellthemachines committed:
Basic mouse interaction

tellthemachines committed:
Try border on container element

tellthemachines committed:
Fix container selected styles.

tellthemachines commented:
@talldan I've addressed all your comments except:

* moving the block visually: attempted it but wasn't happy with the result. Because the block is being moved with the keyboard, the effect is very static, so it looks a bit sad/broken:

<img width="668" alt="Screen Shot 2020-05-28 at 2 53 35 pm" src="https://user-images.githubusercontent.com/8096000/83106577-0ee23880-a100-11ea-8877-e5ffc2bcc0f2.png">

<img width="664" alt="Screen Shot 2020-05-28 at 2 52 53 pm" src="https://user-images.githubusercontent.com/8096000/83106585-1275bf80-a100-11ea-9a67-d4833728cbd0.png">

* Indication of when left/right can be used to nest/unnest: I'm not sure what type of indication would be appropriate here? But as we now have that interaction pattern for navigate mode, perhaps it won't be necessary. 

@enriquesanchez I have added an outline to the parent block, when focus is in an inner block, but when a root block is focused there isn't a good way to do it - it might be a bit weird to outline the whole document. What do you think?

jasmussen commented:
Excited about this PR so I took the liberty of taking it for a spin. It's really impressive, nice work!

![move to](https://user-images.githubusercontent.com/1204802/83113022-54ceda80-a0c7-11ea-9a61-71cbf2411fe9.gif)

The blue line works well, insofar as it appears to use the same style (perhaps even classname? 🤞 ) as should be visible in #21475. 

An indication for the "target nesting container" can work, but I agree the dotted isn't ideal. We probably shouldn't create a new treatment for this actually — can you try simply reusing the same style we use for indicating block focus? I.e. this one:

<img width="1068" alt="Screenshot 2020-05-28 at 09 43 22" src="https://user-images.githubusercontent.com/1204802/83113250-b000cd00-a0c7-11ea-8ceb-c25fc60d9955.png">

The semantics of "focus" are a bit muddy here, since It's not actually what has focus. But it's in the same ballpark — if you press enter, that's the parent container it lands in. I think it can work.

tellthemachines committed:
Update container styles for consistency.

tellthemachines committed:
Fix typo

tellthemachines committed:
Change empty string to null.

tellthemachines commented on packages/block-editor/src/store/actions.js:
```diff
@@ -883,6 +883,18 @@ export function* setNavigationMode( isNavigationMode = true ) {
+ *
+ * @param {string} isBlockMovingMode Enable/Disable block moving mode.
+ */
+export function* setBlockMovingMode( isBlockMovingMode = '' ) {
```
Yeah, I named it after `setNavigationMode` because it mostly works like (and on top of) it. But I needed a way of storing the clientId for the block that initiates moving mode, so I made it take a clientId instead of a boolean. So the naming is awkward, but I'm not sure that `setBlockMovingClientId` is much better, because then we'd end up with `isBlockMovingClientId` which sounds really weird. Am totally open to other suggestions though 😅

draganescu replied:
why not two params then, one for mode `true/false` and one for `clientId`?

afercia commented:
> A quick spin with VoiceOver shows the keyboard flow works, but there are no announcements of the actions taken (that should also be implemented once we decide if this is the way we want to go).

I'd suggest to have a look at the admin Menus screen, where users can move menu items. There's visually hidden text that communicates the item position and level. That might be of inspiration and also improved.

enriquesanchez commented:
I tested again today and it's vastly improving.

Now that the container blocks are highlighted, it's so much easier to understand where the block will be placed.

One thing I caught myself wanting to do and was not possible, was to hit `ESC` at any moment and cancel the whole process, basically a "stop and go back to where it was". Do you think this would be useful to add?

--- 

We also [discussed this PR during the Accessibility Team weekly chat today](https://wordpress.slack.com/archives/C02RP4X03/p1590764828090500). We think this is very promising and had thoughts around how this should be communicated to screen readers. I think it'd be helpful to take a look at the discussion in Slack.

tellthemachines committed:
Announce block moving instructions.

tellthemachines commented:
Thanks for the feedback, @afercia and @enriquesanchez ! I have added a spoken message with instructions on how to move blocks, that should be announced once "Move to" is activated. When users navigate the blocks, each block and its position is announced, exactly the same as in Navigation mode.

So far my testing has had mixed results: 

* NVDA on Firefox and Chrome announces the instructions as soon as "Move to" is pressed.
* VoiceOver announces block name and position first, followed by the instructions. But if it's a Paragraph block, it doesn't announce the instructions at all; instead it reads out the content and announces the font size setting 😮. This is consistent with what happens in Navigation mode, so it might be worth opening a separate issue to look into why Paragraph behaves differently to other blocks with VoiceOver. 

In any case, it would be great to have some more screen reader testing on this!

afercia commented:
Thanks for the update. Testing a bit:

- opening the menu etc. at some point focus is on the "Move to" button:
<img width="998" alt="Screenshot 2020-06-02 at 17 35 03" src="https://user-images.githubusercontent.com/1682452/83544358-0e7ddf00-a4fe-11ea-85ff-0fe7737ff9a6.png">

- on activation, the "Move to" that had focus is removed from the DOM
- at this point, it's not clear to me whether focus is managed programatically or not
- if it's not, I think it must be managed
- focus appears to land on the "breadcrumb" button, which is made "invisible but not unfocusable" in this PR, notice the black outline VoiceOver draws around the (invisible) button:
<img width="997" alt="Screenshot 2020-06-02 at 17 15 57" src="https://user-images.githubusercontent.com/1682452/83544649-877d3680-a4fe-11ea-8ec1-661eeadec663.png">
- I'm not sure focus on an invisible element is ideal
- In any case, I'd recommend to make sure focus is managed programmatically and lands on a meaningful element
- I see the "placeholder line" is just CSS generated content
- maybe consider to use a button instead and move focus there

With screen readers, and generally for accessibility, it is essential to make sure there's always a focused element. The odd Safari + VoiceOver behavior might be in fact triggered by a focus loss.

I have no idea where the VoiceOver announcement "default has been selected" comes from 🙃 

> One thing I caught myself wanting to do and was not possible, was to hit ESC at any moment and cancel the whole process

I'd definitely second this as there should be a way to exit this "move mode" using the keyboard. Also, other modes in Gutenberg (e.g. Edit mode) provide such an exit mechanism via the Escape key.

Lastly, I noticed that a group block can stay in "move mode" when tabbing away from it. I'm not sure this should ever happen. Maybe consider to switch it back to normal mode when moving away from it.

draganescu commented:
> Basic mouse interaction is enabled: clicking on "Move to" and then clicking once on the target block, then pressing Space or Enter, will move the block into place above the target block.

How is this supposed to be when it is not basic anymore? Do we have a mock-up for it?

tellthemachines committed:
Escape out of block moving mode.

tellthemachines commented:
I've added Escape as a shortcut to get out of block moving mode, as per the feedback above.

To address @afercia 's remaining points:

> on activation, the "Move to" that had focus is removed from the DOM
at this point, it's not clear to me whether focus is managed programatically or not

Focus is managed programatically throughout the whole procedure. Block moving uses the same mechanism as Navigation mode, so when "Move to" is removed from the DOM, focus is moved to the breadcrumb button. I've double checked that this works on IE (the browser with the flakiest focus behaviour). 

> I'm not sure focus on an invisible element is ideal - In any case, I'd recommend to make sure focus is managed programmatically and lands on a meaningful element - I see the "placeholder line" is just CSS generated content - maybe consider to use a button instead and move focus there

I initially tried styling the breadcrumb button to act as a placeholder instead, but this doesn't work very well because it's inside a popover, and not a direct child of the block itself, so I couldn't get it to take up the width of the block. I also couldn't decide what to do with the text - for a sighted user it may not be very relevant to have the block name visible. So this was my workaround solution 😅 
I think the breadcrumb is still the best element to use because it has all the information needed about the block and its position on the page. Ideally we'd find a better way to style it than making it invisible; ideas welcome on that front!

> I have no idea where the VoiceOver announcement "default has been selected" comes from 🙃

In the meantime, I found out: it's a bug in the font-size picker in the sidebar. I've done #22815 to address that.

> Lastly, I noticed that a group block can stay in "move mode" when tabbing away from it.

Hmm, I can't reproduce this. Does it stay in "mode mode" after the block is moved, or is there a particular key sequence that causes it?

tellthemachines commented:
> How is this supposed to be when it is not basic anymore? Do we have a mock-up for it?

@draganescu we don't have design input on this so far. The original issue was specifically to address keyboard navigation, but I thought it's awkward to have an action triggered by a clickable button that can't be followed through with a mouse, so I made it work. The actual moving still has to be accomplished by pressing Enter or Space, but you can click on a block to select the placement. It's "basic" because we can probably improve it 😄 but for now the priority should be shipping keyboard functionality.

enriquesanchez commented:
>we don't have design input on this so far

@tellthemachines @draganescu Happy to help here. It's just not clear to me right now what kind of design work this might need. As I understand, [the issue proposes](https://github.com/WordPress/gutenberg/issues/22031) an 'Move to' option in the ellipsis menu. Is there anything else I'm missing?

draganescu commented:
@enriquesanchez I think what's missing is a clear thought on how the mouse/touch interaction will work. For keyboard interaction things are quite ready, as @tellthemachines mentioned, but for mouse it is weird. 

I click on a menu to open it, I click on a menu item, I then suddenly get the mouse to move the item, but to release I have to hit space or enter. It's more of a matter of UX design rather than UI design.

enriquesanchez commented:
@draganescu Thanks for all the context, I wasn't following all the details.

>When this button is in the menu, clicking it is not the same as dragging, but the UI behaves as if drag operation started. While with KBD it works great with the mouse it's not very defined. 

I think there's an opportunity to find some middle ground between the 'Move to' and Drag & drop modes that will benefit a wide range of users, like those who use alternate pointing devices.

While yes you can't drag a block once you've entered the 'Move to' mode, we can still allow the user to point or select where they want the block to be placed. We can show the same blue placement bar as the user hovers over different areas.

So the difference is that you wouldn't be _dragging_, but instead, _clicking_ on where you want the block to be placed. Does that make sense?

Something like this:

![2020-06-10 17-43-52 2020-06-10 17_48_07](https://user-images.githubusercontent.com/3276087/84326478-a012cf80-ab42-11ea-9a93-c51293673ca6.gif)

Or video with commentary, if you prefer: https://www.loom.com/share/ccca8d234bd74483ac9cbb42c5be56db

@tellthemachines @draganescu Does the above example make sense?

draganescu commented:
To me it makes perfect sense! i worry that when you do this for the 1st time, how do you know what you should do? It's a bit too minimalist :D

tellthemachines commented:
> So the difference is that you wouldn't be _dragging_, but instead, _clicking_ on where you want the block to be placed. Does that make sense?

@enriquesanchez that makes sense, yes! I like how it behaves in that mockup. One question: when you click "Move to", in the mockup all the neighbouring blocks get an outline, but currently only the parent block (if the selected block has a parent) is outlined. Wouldn't it be confusing to have parent block _and_ all children outlined at once? 

Whatever we decide, I think that the improvements to mouse behaviour should be addressed as a separate PR, because trying to do it in this one would add considerable complexity to what is already a non-trivial change 🙂

draganescu commented:
I agree @tellthemachines that a followup PR to improve on mouse interaction is best.

Reading the code I am still not convinced by `isBlockMovingMode` returning a `clientId`. Can't we have two things, one that tells us we're in that `mode`, true or false, and one that tells us the `clientId` of the block currently about to be moved?

enriquesanchez commented:
> One question: when you click "Move to", in the mockup all the neighbouring blocks get an outline, but currently only the parent block (if the selected block has a parent) is outlined. Wouldn't it be confusing to have parent block and all children outlined at once?

@tellthemachines Ah, that is a great point that I overlooked. I agree with you, only the parent should get the outline. We could also try doing something similar to https://github.com/WordPress/gutenberg/pull/22508, so make it easier to identify the block structure.

tellthemachines committed:
Fix merge conflict issues.

tellthemachines commented:
> Reading the code I am still not convinced by `isBlockMovingMode` returning a `clientId`. Can't we have two things, one that tells us we're in that `mode`, true or false, and one that tells us the `clientId` of the block currently about to be moved?

Hmm, I'm not too keen on having extra return values as that will add unnecessary complexity to the code. Rather than that, I'll try to think of better naming for these functions.

tellthemachines committed:
Change function names

tellthemachines committed:
Merge branches 'try/keyboard-moving-blocks' and 'master' of github.com:WordPress/gutenberg

tellthemachines committed:
Replace deleted variables.

draganescu commented:
Good progress @tellthemachines One nit I picked is that in both Firefox and Chrome the editor does not scroll to where you are moving focus except after the moving mode is exited by pressing the Enter key:

![no-scroll-move-to](https://user-images.githubusercontent.com/107534/85382473-ae4feb00-b547-11ea-9818-7413094e6400.gif)

tellthemachines commented:
> One nit I picked is that in both Firefox and Chrome the editor does not scroll to where you are moving focus except after the moving mode is exited by pressing the Enter key

That is a navigation mode bug, reported in #22420. It should be fixed separately. I'll likely pick it up soon if no one does it in the meantime 🙂

draganescu reviewed (approved):
This works as expected and offers a great feature for keyboard users. Well have some follow up PR take care of the navigation mode scrolling bug 👏
